# 🏭 Map / Filter / Reduce — Assembly Line ka Trio!

**Folder:** 00_Python → 03_Pythonic_Thinking → 02_Zip_Enumerate_Map_Filter_Reduce
**Level:** 🐔 Inter · **Type:** THEORY+ 💾 Dataset: products_catalog.csv (30 products)

---
## 🎯 Is notebook me kya seekhoge
1. **`map(fn, items)`** = transformer machine — har item pe fn, lazy!
2. **`filter(fn, items)`** = chhalni — sirf passing items
3. **`reduce(fn, items)`** = rolling-pin — sab roll karke EK value (functools import!)
4. Kab map/filter, kab comprehension — the pythonic tussle ⚖️
5. 💾 products: GST pipeline + expensive-filters + inventory totals

**One-liner:**
> assembly line: **map** paints 🎨, **filter** rejects ✂️, **reduce** cake rolls 🌀.

## 🌍 Real-World Hook: Namkeen Factory Assembly Line 🏭

| Machine | Kaam | Code |
|---|---|---|
| 🎨 **Spray-paint station** | HAR item pe same kaam | `map(paint, items)` |
| ✂️ **Quality gate** | Sirf weight-OK rakhi | `filter(ok_weight, items)` |
| 🌀 **Blender end** | Sab daal ke EK mix | `reduce(mix, items)` |

Note: map aur filter **lazy** hain (hello generators! nb2.01 👋) — reduce eager (end-product ek-hi!).

In [1]:
# 🐣 MAP — spray-paint machine 🎨
prices = [100, 250, 799]
def gst(p):
    return round(p * 1.18, 2)

gsted = map(gst, prices)
print("map type   :", type(gsted).__name__, "(lazy machine)")
print("map result :", list(gsted))

# Lambda version — one-liner bina function name ke:
halves = list(map(lambda x: x / 2, prices))
print("map lambda :", halves)

# MAP with TWO iterables — zipper-transform! chain-wise superpower ⚡
qty    = [2, 1, 5]
bills  = list(map(lambda p, q: p * q, prices, qty))
print("map 2-iter :", bills, "← 100*2, 250*1, 799*5 (pei-pe-heist!)")
# Expected OUTPUT:
# map type   : map (lazy machine)
# map result : [118.0, 295.0, 942.82]
# map lambda : [50.0, 125.0, 399.5]
# map 2-iter : [200, 250, 3995] ← 100*2, 250*1, 799*5 (pei-pe-heist!)

map type   : map (lazy machine)
map result : [118.0, 295.0, 942.82]
map lambda : [50.0, 125.0, 399.5]
map 2-iter : [200, 250, 3995] ← 100*2, 250*1, 799*5 (pei-pe-heist!)


In [2]:
# 🐣 FILTER — quality-gate chhalni ✂️
scores = [42, 78, 55, 90, 33, 88]
passed = list(filter(lambda s: s >= 50, scores))
print("filter pass  :", passed)

names = ["apple", "kiwi", "avanade", "mango"]
a_start = list(filter(lambda w: w.startswith("a"), names))
print("filter a-    :", a_start)

# filter lazily + len via sum-1 trick (list nahi banana hota):
count_pass = sum(1 for s in scores if s >= 50)
print("genexpr count:", count_pass, "← filter-list na banao, count chahiye to ye CLASSIC! ⭐")
# Expected OUTPUT:
# filter pass  : [78, 55, 90, 88]
# filter a-    : ['apple', 'avanade']
# genexpr count: 4 ← filter-list na banao, count chahiye to ye CLASSIC! ⭐

filter pass  : [78, 55, 90, 88]
filter a-    : ['apple', 'avanade']
genexpr count: 4 ← filter-list na banao, count chahiye to ye CLASSIC! ⭐


In [3]:
# 🐥 REDUCE — blender-end rolling pin 🌀 (functools se import!)
from functools import reduce

nums = [1, 2, 3, 4]
total = reduce(lambda a, b: a + b, nums)
print("reduce sum   :", total, "← but! sum(nums) pythonic hai:", sum(nums))

prod = reduce(lambda a, b: a * b, nums)
print("reduce prod  :", prod, "← product ke liye math.prod() bhi hai 3.8+")

biggest = reduce(lambda a, b: a if a > b else b, nums)
print("reduce max   :", biggest, "← but max(nums):", max(nums))

# reduce with INITIAL — emergency ki saulekaudit-style:
popularity = reduce(lambda a, b: a + b, [], 0)
print("with initial :", popularity, "(empty pe bhi survives! ✅)")
# Expected OUTPUT:
# reduce sum   : 10 ← but! sum(nums) pythonic hai: 10
# reduce prod  : 24 ← product ke liye math.prod() bhi hai 3.8+
# reduce max   : 4 ← but max(nums): 4
# with initial : 0 (empty pe bhi survives! ✅)

reduce sum   : 10 ← but! sum(nums) pythonic hai: 10
reduce prod  : 24 ← product ke liye math.prod() bhi hai 3.8+
reduce max   : 4 ← but max(nums): 4
with initial : 0 (empty pe bhi survives! ✅)


### ⚖️ Map/Filter vs Comprehension — Pythonic Boxing!

| Scene | Better Choice | Kyun? |
|---|---|---|
| `lambda x: ...` likhna pade | **Comprehension** `[f(x) for x in ...]` | readable, explicit |
| Pehle-se function EXISTS (int, str.upper) | **`map(fn, ...)` clean!** — `list(map(int, s))` | no lambda needed |
| Complex condition+transform mix | **Comprehension** | ek line me filter+map dono |
| Lazy requirement | **genexpr / map-filter chain** — dono lazy | tube afresh |
> Guido ka dil comprehensions pe hai 💛 — map/filter tab jab "function ready hai".

In [4]:
# 💾 REAL DATASET — Products GST Pipeline! 🏭 (30 products)
import csv

def load_products():
    for path in ["../../00_Resources/datasets/products_catalog.csv",
                 "00_Python/00_Resources/datasets/products_catalog.csv"]:
        try:
            with open(path, newline="", encoding="utf-8") as f:
                return list(csv.DictReader(f))
        except FileNotFoundError:
            continue
    raise FileNotFoundError("products_catalog.csv nahi mila!")

products = load_products()
names  = [p["product_name"] for p in products]
prices = [float(p["price"]) for p in products]

# 1) GST map (18%) — first-3 print (spray-paint station!)
gsted = list(map(lambda p: round(p * 1.18, 2), prices))
print("GST first-3     :", list(zip(names, gsted))[:3])

# 2) FILTER: premium (>10,000) — quality gate!
premium = [(n, int(p)) for n, p in zip(names, prices) if p > 10_000]
print("Premium (>10k)  :", premium)

# 3) REDUCE: poora inventory value (price × stock sab ka total!)
from functools import reduce
inv_value = reduce(lambda acc, p: acc + float(p["price"]) * int(p["stock"]), products, 0)
print(f"Inventory value : ₹{inv_value:,.0f}")
# Expected OUTPUT:
# GST first-3     : [('Laptop', 56640.0), ('Smartphone', 17700.0), ('Headphones', 2360.0)]
# Premium (>10k)  : [('Laptop', 48000), ('Smartphone', 15000), ('Tablet', 22000), ('DSLR_Camera', 35000)]
# Inventory value : ₹16,629,090

GST first-3     : [('Laptop', 56640.0), ('Smartphone', 17700.0), ('Headphones', 2360.0)]
Premium (>10k)  : [('Laptop', 48000), ('Smartphone', 15000), ('Tablet', 22000), ('DSLR_Camera', 35000)]
Inventory value : ₹16,629,090


In [5]:
# 💾 Students combo — map/filter/reduce/zip ALL TOGETHER! 🎓🎶
def load_students():
    for path in ["../../00_Resources/datasets/students_marks.csv",
                 "00_Python/00_Resources/datasets/students_marks.csv"]:
        try:
            with open(path, newline="", encoding="utf-8") as f:
                return list(csv.DictReader(f))
        except FileNotFoundError:
            continue
    raise FileNotFoundError("students_marks.csv nahi mila!")

students = load_students()
SUBS = ["math", "science", "english", "hindi", "computer"]

# map: har student ka AVERAGE
avgs = list(map(lambda s: sum(int(s[c]) for c in SUBS) / 5, students))
names = list(map(lambda s: s["name"], students))
print("first-3 avgs :", list(zip(names, avgs))[:3])

# filter: distinction club (avg ≥ 80) 🎖️
dist = list(filter(lambda na: na[1] >= 80, zip(names, avgs)))
print("distinction  :", dist)

# reduce: class total-avg (all avgs ka mean)
class_avg = reduce(lambda a, b: a + b, avgs) / len(avgs)
print("class avg    :", round(class_avg, 2))
print("above-avg    :", sum(1 for a in avgs if a > class_avg), "students above class-normal")
# Expected OUTPUT:
# first-3 avgs : [('Sneha_1', 57.8), ('Pooja_2', 49.8), ('Arjun_3', 69.6)]
# distinction  : [('Ananya_20', 80.2), ('Manish_29', 83.6)]
# class avg    : 65.69
# above-avg    : 27 students above class-normal

first-3 avgs : [('Sneha_1', 57.8), ('Pooja_2', 49.8), ('Arjun_3', 69.6)]
distinction  : [('Ananya_20', 80.2), ('Manish_29', 83.6)]
class avg    : 65.69
above-avg    : 27 students above class-normal


## ⚠️ 5 Common Mistakes (Map/Filter/Reduce Special)

| # | ❌ Galti | ✅ Sahi |
|---|---|---|
| 1 | `reduce` bina import — `NameError` 💥 | `from functools import reduce` PEHLE |
| 2 | `filter(...)` ko index/slice direct | list me wrap: `list(filter(...))` — lazy machine hai! |
| 3 | `reduce` empty list pe (TypeError!) | `reduce(fn, [], initial)` — initial safety cushion |
| 4 | map/filter dobara consume (silent-empty!) — iterator hormones! | list() wrap once-store-ya-fresh regen |
| 5 | `reduce(sum)` kab sum hai | Built-in GO: `sum/max/min/math.prod` — sirf custom-combine ke liye reduce |

> 🎯 **3-second filter:** name-written function ready? → map. Lambda banaunga?
> → comprehension. Ek final single-value? → builtin ya reduce. ✅

## 📌 Recap — 30 Second Me

- `map(fn, items)` lazy transformer (can take MULTI-iterables!) — `list()` to see.
- `filter(fn, items)` lazy gate ✂️ — `sum(1 for ...)` = pythonic counter.
- `reduce(fn, items[, init])` eager blender — import from functools; built-ins first!
- Pythonic rule: ready-fn → map/filter ✔️; custom-lambda → comprehension 💛.
- 💾 proofs: GST map `[Laptop→56640]`, premium ×4, inventory ₹16.6M, class avg 65.69 (27 above 📈).

### ⏭️ Next — `task.ipynb`
7 tasks ⭐→🏆: zip tokens se 💾 **Class Report-Card + Store Console** BOSS tak —
4 machines, 2 datasets, ek mission! 🚀